main.go

```diff
@@ -1076,6 +1086,158 @@ func main() {
 	}, loggedinMiddleware)
 
+	POST("/users/export/{user}", func(w Response, r Request) Output {
+		vars := VARS(r)
+
+		user, err := Q.UserBySlug(r.Context(), vars["user"])
+		if err != nil {
+			return NotFound
+		}
+
+		// export books data
+		books, err := Q.ExportBooksByUserId(r.Context(), user.ID);
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		booksFile, _ := json.MarshalIndent(books, "", "")
+		err = ioutil.WriteFile(BOOK_EXPORT_PATH, booksFile, 0644)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		
+		buf := new(bytes.Buffer)
+		writer := zip.NewWriter(buf)
+
+		bookData, err := ioutil.ReadFile(BOOK_EXPORT_PATH)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		
+		// Write books.json file into the zip file
+		f_book, err := writer.Create(BOOK_EXPORT_PATH)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		_, err = f_book.Write([]byte(bookData))
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+
+
+		//export hightlights data
+		highlights, err := Q.ExportHighlightsByUserId(r.Context(), user.ID);
+		if err != nil {
+			return InternalServerError(err);
+		}
+
+		highlightsFile, _ := json.MarshalIndent(highlights, "", "")
+		_ = ioutil.WriteFile(HIGHLIGHT_EXPORT_PATH, highlightsFile, 0644)
+
+		highlightsData, err := ioutil.ReadFile(HIGHLIGHT_EXPORT_PATH)
+
+		// Write highlights.json file into the zip file
+		f_highlights, err := writer.Create(HIGHLIGHT_EXPORT_PATH)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		_, err = f_highlights.Write([]byte(highlightsData))
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		//export shelves data
+		shelves, err := Q.ExportShelvasByUserId(r.Context(), user.ID)
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		shelvesFile, _ := json.MarshalIndent(shelves, "", "")
+		_ = ioutil.WriteFile(SHELVES_EXPORT_PATH, shelvesFile, 0644)
+		shelvesData, err := ioutil.ReadFile(SHELVES_EXPORT_PATH)
+
+		// Write shelves.json file into the zip file
+		f_shelves, err := writer.Create(SHELVES_EXPORT_PATH)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		_, err = f_shelves.Write([]byte(shelvesData))
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		//export user data
+		_user, err := Q.ExportUserDataByUserId(r.Context(), user.ID)
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		userDataFile, _ := json.MarshalIndent(_user, "", "")
+
+		_ = ioutil.WriteFile(USER_EXPORT_PATH, userDataFile, 0644)
+
+		userData, err := ioutil.ReadFile(USER_EXPORT_PATH)
+
+		// Write user.json file into the zip file
+		f_users, err := writer.Create(USER_EXPORT_PATH)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		_, err = f_users.Write([]byte(userData))
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		err = writer.Close()
+
+		if err != nil {
+			return InternalServerError(err)
+		}
+		w.Header().Set("Content-Type", "application/zip")
```

Comment:
I think you should move writing headers early in this function. as I suggested writing directly to `w` and sending headers should be before sending the body to the client.